cbits/runProcess.c

```diff
@@ -886,78 +854,49 @@ waitForProcess (ProcHandle handle, int *pret)
     return -1;
 }
 
-
+// Returns true on success.
 int
-waitForJobCompletion ( HANDLE hJob, HANDLE ioPort, DWORD timeout, int *pExitCode, setterDef set, getterDef get )
+waitForJobCompletion ( HANDLE hJob )
 {
-    DWORD CompletionCode;
-    ULONG_PTR CompletionKey;
-    LPOVERLAPPED Overlapped;
-    *pExitCode = 0;
-
-    // We have to loop here. It's a blocking call, but
-    // we get notified on each completion event. So if it's
-    // not one we care for we should just block again.
-    // If all processes are finished before this call is made
-    // then the initial call will return false.
-    // List of events we can listen to:
-    // https://msdn.microsoft.com/en-us/library/windows/desktop/ms684141(v=vs.85).aspx
-    while (GetQueuedCompletionStatus (ioPort, &CompletionCode,
-                                      &CompletionKey, &Overlapped, timeout)) {
-
-        // If event wasn't meant of us, keep listening.
-        if ((HANDLE)CompletionKey != hJob)
-          continue;
-
-        switch (CompletionCode)
-        {
-            case JOB_OBJECT_MSG_NEW_PROCESS:
-            {
-                // A new child process is born.
-                // Retrieve and save the process handle from the process id.
-                // We'll need it for later but we can't retrieve it after the
-                // process has exited.
-                DWORD pid    = (DWORD)(uintptr_t)Overlapped;
-                HANDLE pHwnd = OpenProcess(PROCESS_QUERY_LIMITED_INFORMATION, TRUE, pid);
-                set(pid, pHwnd);
-            }
-            break;
-            case JOB_OBJECT_MSG_ABNORMAL_EXIT_PROCESS:
-            case JOB_OBJECT_MSG_EXIT_PROCESS:
-            {
-                // A child process has just exited.
-                // Read exit code, We assume the last process to exit
-                // is the process whose exit code we're interested in.
-                HANDLE pHwnd = get((DWORD)(uintptr_t)Overlapped);
-                if (GetExitCodeProcess(pHwnd, (DWORD *)pExitCode) == 0)
-                {
-                    maperrno();
-                    return 1;
-                }
-
-                // Check to see if the child has actually exited.
-                if (*(DWORD *)pExitCode == STILL_ACTIVE)
-                  waitForProcess ((ProcHandle)pHwnd, pExitCode);
-            }
-            break;
-            case JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO:
-                // All processes in the tree are done.
-                return 0;
-            default:
-                break;
+    JOBOBJECT_BASIC_PROCESS_ID_LIST pid_list;
+    pid_list.NumberOfAssignedProcesses = 1;
+
+    while (true) {
+      // Find a process in the job...
+      bool success = QueryInformationJobObject(
```

Comment:
You need to check this return code. if it's not successful and the return code is `ERROR_MORE_DATA` then you can't trust any of the value in `pid_list`. This is quite under documented but seems to be the case as confirmed on [1][2][3].

This can go wrong when at the time of querying there is more than one process as you've only allocated space for one.
`JOBOBJECT_BASIC_PROCESS_ID_LIST ` is a funky structure, but to allocate space for more you just need to reserve it in the call to `malloc`. i.e. to allocate space for 50 entries you need to reserve 
```
buflen = sizeof(DWORD)*2 + sizeof(ULONG_PTR)*50;
pid_list = (JOBOBJECT_BASIC_PROCESS_ID_LIST *)malloc(buflen);
``` 
because of the change in WoW64 it seems you just have to increase the buffer by an arbitrary amount and try again [2] so choosing a decent sized initial buffer reduces the chance of having to try again. 

[1] https://github.com/mhammond/pywin32/issues/1242
[2] https://social.msdn.microsoft.com/Forums/windowsdesktop/en-US/41a7b8c9-6b5e-4c91-b92d-31310522d0cd/wow64-issue-with-queryinformationjobobject-and-jobobjectbasicprocessidlist-including-windows-10?forum=windowssdk
[3] https://www.winehq.org/pipermail/wine-cvs/2015-March/105916.html

Reply:
Wow, indeed this quite underdocumented. I had assumed that I could specify a quantity less than the actual process count and it would simply give me a subset of processes. Moreover, this appears to even work as-is.

Nevertheless, I'll rework this.